### PR TITLE
Add DRep and CC key types

### PR DESCRIPTION
**Pull request by disassembler**
dependent on #231

### CODE REVIEW

File: cabal.project [hunk 1]
```diff
@@ -1,75 +1,60 @@
...
+  root-keys:
+    3e0cce471cf09815f930210f7827266fd09045445d65923e6d0238a6cd15126f
+    443abb7fb497a134c343faf52f0b659bd7999bc06b7f63fa76dc99d631f9bea1
+    a86a1f6ce86c449c46666bda44268677abf29b5b2d2eb5ec7af903ec2f117a82
+    bcec67e8e99cabfa7764d75ad9b158d72bfacf70ca1d0ec8bc6b4406d1bf8413
+    c00aae8461a256275598500ea0e187588c35a5d5d7454fb57eac18d9edb86a56
+    d4a35cd3121aa00d18544bb0ac01c3e1691d618f462c46129271bccf39f7e8ee
+
+-- See CONTRIBUTING for information about these, including some Nix commands
+-- you need to run if you change them
```
Comment (Crypto2099) [inline cabal.project:14]: Is this legacy code from another source? Do we need to update CONTRIBUTING with relevant details?

File: command-line/lib/Command/Key/WalletId.hs [hunk 2]
```diff
@@ -57,8 +59,8 @@ mod :: (Cmd -> parent) -> Mod CommandFields parent
 mod liftCmd = command "walletid" $
     info (helper <*> fmap liftCmd parser) $ mempty
         <> progDesc "Shows the cardano-wallet wallet ID for a given key"
-        <> footerDoc (Just $ string $ mconcat
-            [ "A wallet ID is a 40-digit hexadecimal string derived "
+        <> footerDoc (Just $ pretty $ mconcat
+            [ "A wallet ID is a 40-digit hexadecimal pretty derived " :: Text
```
Comment (Crypto2099) [inline command-line/lib/Command/Key/WalletId.hs:63]: This looks like a situation where "find + replace" found a legitimate use of the word "string" and replaced it with pretty even though that's not desired behavior! :D

```suggestion
            [ "A wallet ID is a 40-digit hexadecimal string derived " :: Text
```

**Review by Crypto2099 — COMMENTED** [review]
Couple of minor things noticed but overall looking like good progress

**Comment by Crypto2099** [comment]
cc: @aleeusgr please see changes here and would love your feedback on the Nix build issues we're still seeing!

File: core/lib/Cardano/Address/Style/Shelley.hs [hunk 3]
```diff
@@ -370,6 +382,60 @@ deriveDelegationPrivateKey accXPrv =
...
+-- software are by convention only using one delegation key per account, and always
+-- the first account (with index 0').
+--
+-- Deriving DRep keys for something else than the initial account is not
+-- recommended and can lead to incompatibility with existing wallet softwares
+-- (Daedalus, Yoroi, Adalite...).
+--
+deriveDRepPrivateKey
+    :: Shelley 'AccountK XPrv
+    -> Shelley 'DelegationK XPrv
```
Comment (paweljakubas) [inline core/lib/Cardano/Address/Style/Shelley.hs:397]: we should have here `'DRepK`

File: core/lib/Cardano/Address/Style/Shelley.hs [hunk 4]
```diff
@@ -370,6 +382,60 @@ deriveDelegationPrivateKey accXPrv =
...
+-- software are by convention only using one delegation key per account, and always
+-- the first account (with index 0').
+--
+-- Deriving CCCold keys for something else than the initial account is not
+-- recommended and can lead to incompatibility with existing wallet softwares
+-- (Daedalus, Yoroi, Adalite...).
+--
+deriveCCColdPrivateKey
+    :: Shelley 'AccountK XPrv
+    -> Shelley 'DelegationK XPrv
```
Comment (paweljakubas) [inline core/lib/Cardano/Address/Style/Shelley.hs:415]: we should have here `'CCColdK`

File: core/lib/Cardano/Address/Style/Shelley.hs [hunk 5]
```diff
@@ -370,6 +382,60 @@ deriveDelegationPrivateKey accXPrv =
...
+-- software are by convention only using one delegation key per account, and always
+-- the first account (with index 0').
+--
+-- Deriving CCHot keys for something else than the initial account is not
+-- recommended and can lead to incompatibility with existing wallet softwares
+-- (Daedalus, Yoroi, Adalite...).
+--
+deriveCCHotPrivateKey
+    :: Shelley 'AccountK XPrv
+    -> Shelley 'DelegationK XPrv
```
Comment (paweljakubas) [inline core/lib/Cardano/Address/Style/Shelley.hs:433]: we should have here `'CCHotK`

**Review by paweljakubas — CHANGES_REQUESTED** [review]
So we have somewhere decided to use the following derivation paths
`1852H/1815H/0H/3/0` for DRep
`1852H/1815H/0H/4/0` for CCCold
`1852H/1815H/0H/5/0` for CCHot
Correct?

Also it would be good to adjust https://github.com/IntersectMBO/cardano-addresses/blob/master/core/test/Cardano/Address/Style/ShelleySpec.hs
(extend role generator and add golden here)
Finally, cli support would need to be added (of course could be in separate PR)

**Comment by paweljakubas** [comment]
cherry-picked the commit and finished here https://github.com/IntersectMBO/cardano-addresses/pull/259

**Pull request closed without merging**